datastore/keycompatibility.go

```diff
@@ -0,0 +1,67 @@
+// Copyright 2011 Google Inc. All rights reserved.
+// Use of this source code is governed by the Apache 2.0
+// license that can be found in the LICENSE file.
+
+// This change adds additional compatibility to help customers
+// transition from google.golang.org/appengine/datastore (oldds) to cloud.google.com/go/datastore (newds).
+// Each lib (oldds and newds) contain functions Key.Encode() and Key.Decode(). These functions
+// create base64 representations of a json marshalled type of datastore keys.  Customers have been using
+// these encoded values to communicate between services in appengine.  Protobuf key types
+// change between oldds and newds making the corresponding base64 key strings incompatible.
+// Customer who attempt to upgrade to newds that use this pattern will fail.
+// keycompatibility.go placed in oldds enables forward compatibility of newds encoded keys.
+// An update to newds will also be necessary to enable backward compatibility.
+
+package datastore
+
+import (
+	"errors"
+
+	nds "cloud.google.com/go/datastore"
+	"golang.org/x/net/context"
+	"google.golang.org/appengine/internal"
+)
+
+var errKeyConversion string = `Key conversions must be enabled in the application.\n  
+See https://github.com/golang/appengine#key-encode-decode-compatibiltiy-to -help-with-datastore-library-migrations for more details.`
+
+var convKey *keyConverter
+
+// EnableKeyConversion enables forward key conversion abilities.  Calling this function in a single handler will enable
+// the feature for all handlers.  This function can be called in the /_ah/start handler.  Support for key converstion.
+// Variable holds the appid so that key conversion can retrieve it without a context.
+func EnableKeyConversion(ctx context.Context) {
+	convKey = &keyConverter{
+		appid: internal.FullyQualifiedAppID(ctx),
```

Comment:
This feels bad.

Did you consider adding an "appID" or "projectID" (or just the context, which contains the app id) argument to another function along the lines of `DecodeKey`?

`DecodeKeyContext(ctx context.Context, encoded string) (*Key, error)`

Reply:
The implementation feels bad (I agree) but it does allow for the least lines of code change on the customers part.  If we add a new function the customer will need to explore their code to find all the places they need to use DecodeKeyContext(ctx context.Context, encoded string) (*Key, error) and will also need scope on the context.  With this implementation the customer just needs to call EnableKeyConversion(ctx) in the startup handler where we have a context.  While not pretty,  I thought it was a clever way to use Go Errors to minimize code changes on the customers part.

Reply:
Agreed upon implementation has been submitted